types/react-redux/index.d.ts

```diff
@@ -9,22 +9,23 @@ import * as Redux from 'redux';
 
 type ComponentClass<P> = React.ComponentClass<P>;
 type StatelessComponent<P> = React.StatelessComponent<P>;
+type Component<P> = ComponentClass<P> | StatelessComponent<P>;
 type ReactNode = React.ReactNode;
 type Store<S> = Redux.Store<S>;
 type Dispatch<S> = Redux.Dispatch<S>;
 type ActionCreator<A> = Redux.ActionCreator<A>;
 
-interface ComponentDecorator<TOriginalProps, TOwnProps> {
-    (component: ComponentClass<TOriginalProps> | StatelessComponent<TOriginalProps>): ComponentClass<TOwnProps>;
+interface ComponentDecorator<TStateProps, TDispatchProps, TOwnProps> {
+    (component: Component<TStateProps & TDispatchProps & TOwnProps>): ComponentClass<TOwnProps>;
 }
 
 /**
  * Decorator that infers the type from the original component
  *
  * Can't use the above decorator because it would default the type to {}
  */
-export interface InferableComponentDecorator {
-    <P, TComponentConstruct extends (ComponentClass<P> | StatelessComponent<P>)>(component: TComponentConstruct): TComponentConstruct;
+export interface InferableComponentDecorator<IOwnProps> {
```

Comment:
Does this still work as intended, i.e., does it properly infer `IOwnProps`? Also, please rename it to `TOwnProps` for consistency.